backend/src/models/User.ts

```diff
@@ -0,0 +1,328 @@
+  @Column({
+    allowNull: false
+  })
+  email: string;
+
+  @Column({
+    allowNull: false
+  })
+  username: string;
+
+  @Column
+  hash: string;
+
+  @Column
+  salt: string;
+
+  @Column
+  activationKey: string;
+
+  @Column
+  resetPasswordKey: string;
+
+  @Column
+  verified: boolean;
+
+  setPassword (password, cb) {
+    if (!password) {
+      return cb(new Error(options.missingPasswordError));
+    }
+
+    const self = this;
+
+    crypto.randomBytes(options.saltlen, function (err, buf) {
+      if (err) {
+        return cb(err);
+      }
+
+      const salt = buf.toString('hex');
+
+      crypto.pbkdf2(password, salt, options.iterations, options.keylen, options.digest, function (err, hashRaw) {
+        if (err) {
+          return cb(err);
+        }
+
+        // TODO: type
+        self.set(options.hashField, new Buffer(hashRaw as any as string, 'binary').toString('hex'));
+        self.set(options.saltField, salt);
+
+        cb(null, self);
+      });
+    });
+  }
+
+  setActivationKey (cb) {
+
+    const self = this;
+
+    if (!options.activationRequired) {
+      return cb(null, self);
+    }
+
+    crypto.randomBytes(options.activationkeylen, function (err, buf) {
+      if (err) {
+        return cb(err);
+      }
+
+      const randomHex = buf.toString('hex');
+      self.set(options.activationKeyField, randomHex);
+      cb(null, self);
+    });
+  }
+
+  authenticate (password, cb) {
+    const self = this;
+
+    // prevent to throw error from crypto.pbkdf2
+    if (!this.get(options.saltField)) {
+      return cb(null, false, { message: options.noSaltValueStoredError });
+    }
+
+    crypto.pbkdf2(password, this.get(options.saltField), options.iterations, options.keylen, options.digest, function (err, hashRaw) {
+      if (err) {
+        return cb(err);
+      }
+
+      // TODO types
+      const hash = new Buffer(hashRaw as any as string, 'binary').toString('hex');
+
+      if (hash === self.get(options.hashField)) {
+        return cb(null, self);
+      } else {
+        return cb(null, false, { message: options.incorrectPasswordError });
+      }
+    });
+  }
+
+  static authenticate () {
+    const self = User;
+    return function (username, password, cb) {
+      self.findByUsername(username, function (err, user) {
+        if (err) { return cb(err); }
+
+        if (user) {
+          return user.authenticate(password, cb);
+        } else {
+          return cb(null, false, { message: options.incorrectUsernameError });
+        }
+      });
+    };
+  }
+
+  static serializeUser () {
+    return function (user, cb) {
+      cb(null, user.get(options.usernameField));
+    };
+  }
+
+  static deserializeUser () {
+    const self = User;
+    return function (username, cb) {
+      self.findByUsername(username, cb);
+    };
+  }
+
+  static register (user, password, cb) {
+    const self = User,
+      fields = {};
+    // TODO: check this
+    if (user instanceof User) {
```

Comment:
I would put all these User checks inside a function with a name like `user = this.getUserInstance(payload)`.